File: ICDB/src/main/java/cipher/Key.java
package cipher;

import cipher.signature.Convert;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import org.bouncycastle.crypto.AsymmetricCipherKeyPair;
import org.bouncycastle.crypto.params.AsymmetricKeyParameter;
import org.bouncycastle.crypto.params.KeyParameter;
import org.bouncycastle.crypto.util.PrivateKeyFactory;
import org.bouncycastle.crypto.util.PublicKeyFactory;
import org.bouncycastle.jce.provider.BouncyCastleProvider;
import org.bouncycastle.openssl.PEMKeyPair;
import org.bouncycastle.openssl.PEMParser;
import org.bouncycastle.openssl.jcajce.JcaPEMKeyConverter;

import java.io.FileReader;
import java.io.IOException;
import java.io.Reader;
import java.security.KeyPair;
import java.security.Security;

/**
 * <p>
 * </p>
 * Created on 7/13/2016
 *
 * @author Dan Kondratyuk
 */
public class Key {

    private final AsymmetricCipherKeyPair rsaKeyPair;
    private final byte[] rawMacKey;
    private final KeyParameter macKey;

    private static final Logger logger = LogManager.getLogger();

    static {
        Security.addProvider(new BouncyCastleProvider());
    }

    public Key(String macKey, String rsaKeyFile) {
        this.rawMacKey = Convert.fromBase64(macKey);
        this.macKey = new KeyParameter(rawMacKey);
        this.rsaKeyPair = readKeys(rsaKeyFile);
    }

    public byte[] getRawMacKey() {
        return rawMacKey;
    }

    public KeyParameter getMacKey() {
        return macKey;
    }

    public AsymmetricKeyParameter getPublic() {
        return rsaKeyPair.getPublic();
    }

    public AsymmetricKeyParameter getPrivate() {
        return rsaKeyPair.getPrivate();
    }

    private static AsymmetricCipherKeyPair readKeys(final String rsaKeyFile)
    {
        try (Reader reader = new FileReader(rsaKeyFile)) {
            PEMParser parser = new PEMParser(reader);

            Object o;
            while ((o = parser.readObject()) != null)
            {
                if (o instanceof PEMKeyPair)
                {
                    PEMKeyPair keyPair = (PEMKeyPair) o;

                    KeyPair pair = new JcaPEMKeyConverter()
                            .setProvider("BC")
                            .getKeyPair(keyPair);

                    AsymmetricKeyParameter privateKey = PrivateKeyFactory.createKey(pair.getPrivate().getEncoded());
                    AsymmetricKeyParameter publicKey = PublicKeyFactory.createKey(pair.getPublic().getEncoded());

                    return new AsymmetricCipherKeyPair(privateKey, publicKey);
                }
            }

            return (AsymmetricCipherKeyPair) parser.readObject();
        } catch (IOException e) {
            logger.error("Failed to read RSA key file: {}", e.getMessage());
            System.exit(1);
        }

        return null;
    }
}

File: ICDB/src/main/java/crypto/Key.kt
package crypto

import org.bouncycastle.crypto.AsymmetricCipherKeyPair
import org.bouncycastle.crypto.params.AsymmetricKeyParameter
import org.bouncycastle.crypto.params.KeyParameter
import org.bouncycastle.crypto.util.PrivateKeyFactory
import org.bouncycastle.crypto.util.PublicKeyFactory
import org.bouncycastle.jce.provider.BouncyCastleProvider
import org.bouncycastle.openssl.PEMKeyPair
import org.bouncycastle.openssl.PEMParser
import org.bouncycastle.openssl.jcajce.JcaPEMKeyConverter

import java.io.FileReader
import java.math.BigInteger
import java.security.KeyFactory
import java.security.Security
import java.security.spec.RSAPublicKeySpec

/**
 * Wraps RSA and MAC keys for easy access
 *
 * Created on 7/13/2016
 * @author Dan Kondratyuk
 */
class Key(macKey: String, rsaKeyFile: String) {

    init {
        Security.addProvider(BouncyCastleProvider())
    }

    val rawMacKey = Convert.fromBase64(macKey)
    val macKey = KeyParameter(rawMacKey)
    var modulus: BigInteger =  BigInteger("1")
    var exponent: BigInteger =  BigInteger("1")
    private val rsaKeyPair = readRSAKeys(rsaKeyFile)

    val publicRsaKey: AsymmetricKeyParameter = rsaKeyPair.public
    val privateRsaKey: AsymmetricKeyParameter = rsaKeyPair.private

    private fun readRSAKeys(rsaKeyFile: String): AsymmetricCipherKeyPair {
        FileReader(rsaKeyFile).use { reader ->
            val parser = PEMParser(reader)

            var o: Any? = null
            while ({ o = parser.readObject(); o }() != null) {
                if (o is PEMKeyPair) {
                    val pair = JcaPEMKeyConverter()
                        .setProvider("BC")
                        .getKeyPair(o as PEMKeyPair)

                    val privateKey = PrivateKeyFactory.createKey(pair.private.encoded)
                    val publicKey = PublicKeyFactory.createKey(pair.public.encoded)

                    val keyFac = KeyFactory.getInstance("RSA")
                    val pkSpec = keyFac.getKeySpec(pair.public,
                            RSAPublicKeySpec::class.java)

                    modulus = pkSpec.modulus
                    exponent=pkSpec.publicExponent

                    return AsymmetricCipherKeyPair(privateKey, publicKey)
                }
            }

            return parser.readObject() as AsymmetricCipherKeyPair
        }
    }

}
